In [301]:
import os
import numpy as np
import pandas as pd
import pickle
import json

In [302]:
song_path1="Data/九宫大成-对齐2/"
song_path2="Data/昆曲剧本+简谱/"
lmd_output_path="Data/GeneratedNpy/"
songmass_output_path="Data/GeneratedNpy/"
songList=[]
songOrigin=[]
songData=[]
noteList=[]
noteOrigin=[]
nameData=[]

In [303]:
def readMusic(path):
    with open(path,'r',encoding="utf-8") as f:
        return json.load(f)

In [304]:
files = os.listdir(song_path1)
for chapter in files:
    if os.path.isdir(os.path.join(song_path1,chapter)):
        songs=os.listdir(os.path.join(song_path1,chapter))
        for song in songs:
            nameData.append(song)
            songData.append(readMusic(os.path.join(song_path1,chapter,song)))
        songList.extend(songs)
        songOrigin.append(songs)
        
print(f'总卷数：{len(songOrigin)}')
print(f'歌曲数：{len(songList)}')
print(f'歌曲数：{len(songData)}')

总卷数：13
歌曲数：725
歌曲数：725


In [322]:
nameData

['卷五十一_001_太师令_琵琶记.json',
 '卷五十一_002_又一体_满牀笏.json',
 '卷五十一_003_太师垂绣带_卧冰.json',
 '卷五十一_004_又一体_散曲.json',
 '卷五十一_005_太师醉腰图_梦花酣.json',
 '卷五十一_006_太师入锁恋_鸳鸯棒.json',
 '卷五十一_007_太师见学士_散曲.json',
 '卷五十一_008_阮二郎_蜃中楼.json',
 '卷五十一_010_六奏新音_散曲.json',
 '卷五十一_011_梁州新郎_琵琶记.json',
 '卷五十一_013_又一体_惊鸿记.json',
 '卷五十一_014_梁州锦序_蝴蝶梦.json',
 '卷五十一_015_梁州四集_散曲.json',
 '卷五十一_016_柰子宜春_十孝记.json',
 '卷五十一_019_柰子乐_灌圆记.json',
 '卷五十一_020_又一体_千金记.json',
 '卷五十一_028_学士解溪沙_散曲.json',
 '卷五十一_029_学士醉江风_散曲.json',
 '卷五十一_030_宜春乐_怜香伴.json',
 '卷五十一_032_又一体_永团圆.json',
 '卷五十一_033_春色.json',
 '卷五十一_034_三十腔_散曲.json',
 '卷五十一_035_浣溪令_牡丹亭.json',
 '卷五十一_036_浣溪莲_散曲.json',
 '卷五十一_037_浣溪箱_一诺媒.json',
 '卷五十一_038_浣溪乐_散曲.json',
 '卷五十一_039_浣溪天乐_散曲.json',
 '卷五十一_040_浣溪帽_西园记.json',
 '卷五十一_041_溪沙四集_散曲.json',
 '卷五十一_042_浣溪刘红莲_绿牡丹.json',
 '卷五十一_043_筛筛金莲_一诺媒.json',
 '卷五十一_044_澄帽落东瓯_白玉楼.json',
 '卷五十一_045_水澄帽_散曲.json',
 '卷五十一_047_香南枝_水浒记.json',
 '卷五十一_048_香娇枝_水浒记.json',
 '卷五十一_049_大看灯_金明池.json',
 '卷五十一_050_迓鼓娘_梦花酣.json',
 '卷五十七_001_二郎神_太平园.json',
 '卷五十七

In [306]:
songData[0]

{'key': 'D',
 'note1': {'note': 'C5',
  'rhythm': 0.25,
  'haspot': False,
  'gongchi': '仩',
  'word': '細'},
 'note2': {'note': 'A5',
  'rhythm': 0.25,
  'haspot': False,
  'gongchi': '五',
  'word': '端'},
 'note3': {'note': 'G4',
  'rhythm': 0.25,
  'haspot': False,
  'gongchi': '六',
  'word': '相'},
 'note4': {'note': 'A5',
  'rhythm': 0.25,
  'haspot': False,
  'gongchi': '五',
  'word': ''},
 'note5': {'note': 'C5',
  'rhythm': 0.25,
  'haspot': False,
  'gongchi': '仩(?)',
  'word': '這'},
 'note6': {'note': 'G4',
  'rhythm': 0.25,
  'haspot': False,
  'gongchi': '工',
  'word': '是'},
 'note7': {'note': 'A5',
  'rhythm': 0.25,
  'haspot': True,
  'gongchi': '尺',
  'word': ''},
 'note8': {'note': 'G4',
  'rhythm': 0.0625,
  'haspot': False,
  'gongchi': '六',
  'word': ''},
 'note9': {'note': 'E4',
  'rhythm': 0.125,
  'haspot': False,
  'gongchi': '仩',
  'word': ''},
 'note10': {'note': 'D4',
  'rhythm': 0.125,
  'haspot': False,
  'gongchi': '五',
  'word': ''},
 'note11': {'note': '',
 

In [307]:
for song in songData:  
    ls=[]
    sent=[]
    for key,value in song.items():
        if 'note' in key:
            sent.append(value)
        elif 'juhao' in key:
            ls.append(sent)
            noteList.append(sent)
            sent=[]
    noteOrigin.append(ls)

In [308]:
len(noteOrigin),len(noteList),noteOrigin[0][0],noteList[0]

(725,
 6516,
 [{'note': 'C5', 'rhythm': 0.25, 'haspot': False, 'gongchi': '仩', 'word': '細'},
  {'note': 'A5', 'rhythm': 0.25, 'haspot': False, 'gongchi': '五', 'word': '端'},
  {'note': 'G4', 'rhythm': 0.25, 'haspot': False, 'gongchi': '六', 'word': '相'},
  {'note': 'A5', 'rhythm': 0.25, 'haspot': False, 'gongchi': '五', 'word': ''},
  {'note': 'C5',
   'rhythm': 0.25,
   'haspot': False,
   'gongchi': '仩(?)',
   'word': '這'},
  {'note': 'G4', 'rhythm': 0.25, 'haspot': False, 'gongchi': '工', 'word': '是'},
  {'note': 'A5', 'rhythm': 0.25, 'haspot': True, 'gongchi': '尺', 'word': ''},
  {'note': 'G4',
   'rhythm': 0.0625,
   'haspot': False,
   'gongchi': '六',
   'word': ''},
  {'note': 'E4', 'rhythm': 0.125, 'haspot': False, 'gongchi': '仩', 'word': ''},
  {'note': 'D4', 'rhythm': 0.125, 'haspot': False, 'gongchi': '五', 'word': ''},
  {'note': '', 'rhythm': '', 'haspot': False, 'gongchi': '六(□)', 'word': ''},
  {'note': 'C4', 'rhythm': 0.25, 'haspot': False, 'gongchi': '上', 'word': '誰'},
  {'

In [309]:
noteOrigin[1][0]

[{'note': '', 'rhythm': '', 'haspot': False, 'gongchi': '', 'word': '酒'},
 {'note': 'G4', 'rhythm': 0.0625, 'haspot': False, 'gongchi': '六', 'word': ''},
 {'note': 'E4', 'rhythm': 0.0625, 'haspot': False, 'gongchi': '工', 'word': ''},
 {'note': 'G4', 'rhythm': 0.25, 'haspot': False, 'gongchi': '六', 'word': '星'},
 {'note': 'E4',
  'rhythm': 0.125,
  'haspot': False,
  'gongchi': '工(□)',
  'word': '何'},
 {'note': 'G4', 'rhythm': 0.125, 'haspot': False, 'gongchi': '六', 'word': ''},
 {'note': 'A5',
  'rhythm': 0.0625,
  'haspot': False,
  'gongchi': '一',
  'word': '事'},
 {'note': 'G4', 'rhythm': 0.0625, 'haspot': False, 'gongchi': '五', 'word': ''},
 {'note': 'E4', 'rhythm': 0.125, 'haspot': False, 'gongchi': '六', 'word': ''},
 {'note': '', 'rhythm': '', 'haspot': False, 'gongchi': '工', 'word': ''},
 {'note': 'A4',
  'rhythm': 0.125,
  'haspot': True,
  'gongchi': '四(丶)',
  'word': '明'},
 {'note': 'C4', 'rhythm': 0.0625, 'haspot': False, 'gongchi': '尺', 'word': ''},
 {'note': 'D4', 'rhythm':

In [310]:
pitch2intDict={
    "C":0,
    "D":2,
    "E":4,
    "F":5,
    "G":7,
    "A":9,
    "B":11
}
def pitch2int(pitch):
    res=0
    if pitch=="#":
        return 0
    for s in pitch:
        if s in pitch2intDict.keys():
            res+=pitch2intDict.get(s)
        else:
            res+=int(s)*12
    return res

In [311]:
def getInfo(obj:list):
    wordLs=[]
    wordOrigin=[]
    midiLs=[]
    midiOrigin=[]
    syll=[]
    midi=[]
    rest=0.0
    for i in range(len(obj)):
        word=obj[i]["word"]
        pitch=obj[i]["note"]
        rhy=obj[i]["rhythm"]
        if rhy=="":
            rhy=0
        isPot=obj[i]["haspot"]
        if not word and (pitch or rhy):
            if pitch=='#':
                rest=rhy*rhyFactor
            else:
                if isPot:
                    midi.append([pitch2int(pitch),rhy*1.5*rhyFactor,rest])
                else:
                    midi.append([pitch2int(pitch),rhy*rhyFactor,rest])
                rest=0.0
                # print(syll+[i])
                word=syll[0]
                syll.append(word)
        elif not word and not(pitch or rhy):
            if isPot:
                midi.append([midi[-1][0],midi[-1][1]/3,rest])
                midi[-2][1]=midi[-2][1]*2/3
            else:
                if type(midi[-1][1])==str:
                    print(midi[-1])
                    print(syll[-1])
                midi.append([midi[-1][0],midi[-1][1]/2,rest])
                midi[-2][1]=midi[-2][1]/2
            rest=0.0
            # print(syll+[i])
            word=syll[0]
            syll.append(word)
        else:
            if len(syll)>1:
                wordOrigin.append(syll)
                midiOrigin.append(midi)
                wordLs.extend(syll)
                midiLs.extend(midi)
                # print(syll+[i])
                syll=[]
                midi=[]
            elif len(syll)==1:
                wordOrigin.append(syll)
                midiOrigin.append(midi)
                wordLs.extend(syll)
                midiLs.extend(midi)
                syll=[]
                midi=[]
            syll.append(word)
            if isPot:
                midi.append([pitch2int(pitch),rhy*1.5*rhyFactor,rest])
            else:
                midi.append([pitch2int(pitch),rhy*rhyFactor,rest])
            rest=0.0
            # print(syll+[i])
            
    return wordLs,wordOrigin,midiLs,midiOrigin

In [312]:
word,syll,midils,midiOri=getInfo(noteOrigin[0][0])

In [313]:
syll

[['細'],
 ['端'],
 ['相', '相'],
 ['這'],
 ['是', '是', '是', '是', '是', '是'],
 ['誰', '誰', '誰', '誰'],
 ['筆', '筆', '筆', '筆', '筆'],
 ['仗', '仗', '仗', '仗'],
 ['覷', '覷', '覷', '覷', '覷', '覷'],
 ['着'],
 ['他', '他', '他'],
 ['心', '心', '心'],
 ['兒', '兒', '兒', '兒', '兒', '兒'],
 ['感', '感'],
 ['傷', '傷', '傷', '傷'],
 ['好', '好'],
 ['似', '似', '似'],
 ['我', '我', '我'],
 ['雙', '雙'],
 ['親', '親', '親', '親', '親'],
 ['摸', '摸', '摸'],
 ['樣', '樣', '樣', '樣', '樣', '樣', '樣', '樣', '樣', '樣', '樣', '樣'],
 ['怎'],
 ['穿'],
 ['着', '着', '着', '着'],
 ['破', '破', '破', '破', '破'],
 ['損', '損', '損'],
 ['衣', '衣', '衣', '衣', '衣', '衣', '衣'],
 ['裳', '裳'],
 ['道', '道', '道', '道'],
 ['別', '別', '別'],
 ['後', '後', '後', '後'],
 ['容', '容', '容'],
 ['顏', '顏', '顏', '顏', '顏', '顏'],
 ['無', '無', '無'],
 ['恙', '恙', '恙', '恙', '恙', '恙', '恙', '恙', '恙', '恙', '恙', '恙'],
 ['怎', '怎', '怎', '怎', '怎'],
 ['這', '這', '這', '這'],
 ['般', '般', '般', '般', '般'],
 ['凄', '凄', '凄'],
 ['凄', '凄', '凄'],
 ['形', '形', '形', '形'],
 ['狀', '狀', '狀', '狀'],
 ['誰', '誰', '誰', '誰'],
 ['來', '來'],
 ['往', '往'

In [314]:
midiOri

[[[60, 1.0, 0.0]],
 [[69, 1.0, 0.0]],
 [[55, 1.0, 0.0], [69, 1.0, 0.0]],
 [[60, 1.0, 0.0]],
 [[55, 1.0, 0.0],
  [69, 1.5, 0.0],
  [55, 0.25, 0.0],
  [52, 0.5, 0.0],
  [50, 0.25, 0.0],
  [50, 0.25, 0.0]],
 [[48, 1.0, 0.0], [50, 1.0, 0.0], [52, 1.5, 0.0], [50, 0.5, 0.0]],
 [[48, 1.0, 0.0],
  [50, 1.0, 0.0],
  [48, 1.0, 0.0],
  [57, 0.5, 0.0],
  [43, 0.5, 0.0]],
 [[57, 0.5, 0.0], [50, 0.25, 0.0], [48, 0.25, 0.0], [57, 1.0, 0.0]],
 [[69, 0.5, 0.0],
  [69, 0.25, 0.0],
  [55, 0.25, 0.0],
  [52, 0.5, 0.0],
  [50, 0.25, 0.0],
  [48, 0.25, 0.0]],
 [[57, 2.0, 0.0]],
 [[48, 1.0, 0.0], [48, 0.5, 0.0], [57, 0.5, 0.0]],
 [[48, 1.0, 0.0], [50, 0.75, 0.0], [48, 0.25, 0.0]],
 [[57, 1.0, 0.0],
  [48, 1.0, 0.0],
  [50, 1.5, 0.0],
  [48, 0.5, 0.0],
  [57, 1.0, 0.0],
  [43, 1.0, 0.0]],
 [[57, 4.0, 0.0], [48, 1.0, 0.0]],
 [[50, 1.5, 0.0], [52, 0.5, 0.0], [50, 2.0, 0.0], [50, 2.0, 0.0]],
 [[50, 1.0, 0.0], [52, 1.0, 0.0]],
 [[52, 0.5, 0.0], [55, 0.25, 0.0], [52, 0.25, 0.0]],
 [[50, 1.0, 0.0], [52, 1.0, 0.0], 

In [315]:
wordList=[]
wordOrigin=[]
midiList=[]
midiOrigin=[]
rhyFactor=4

In [316]:
for songNote in noteOrigin:
    sentWordLs=[]
    sentWordOri=[]
    sentMidiLs=[]
    sentMidiOri=[]
    for sent in songNote:
        word,syll,midils,midiori=getInfo(sent)
        sentWordLs.extend(word)
        sentWordOri.append(syll)
        sentMidiLs.extend(midils)
        sentMidiOri.append(midiori)
    wordList.extend(sentWordLs)
    wordOrigin.append(sentWordOri)
    midiList.extend(sentMidiLs)
    midiOrigin.append(sentMidiOri)

In [317]:
len(wordOrigin),len(wordList),len(midiOrigin),len(midiList)

(725, 65859, 725, 65859)

In [318]:
len(wordOrigin[1]),len(wordOrigin[1][0])

(8, 6)

### 生成说明
最终生成了lmd相似的那种格式的数据，以wordOrigin为例说明
- 第一维代表每个json文件也就是每个歌曲，长度725符合
- 第二维代表一句话，一juhao为分割
- 第三维代表这首歌中一个字，他是从音开始到结束，在json文件中一个字可以对应几个音，我看note中word有时候为“”说明就是这种情况，在这时将相同的字加入到数组中，为什么我要加入相同的字而不是像word说明里说的标注为BIE呢，因为那个songmass的程序他选择的是这种类型，就是一个字多音时则将相同的字重复加入数组里，可以去看<u>https://github.com/microsoft/muzic/blob/main/songmass/data/generate_lmd_dataset.py</u> 的第109行，他选择的是data的1和2，然后他选择的数据是Lyrics-Conditioned-Neural-Melody-Generation/lmd-full_MIDI_dataset/Sentence_and_Word_Parsing 这个文件夹里的，这里的npy文件我看了，第1和2就是一个是音高数组一个是字，这个字如果对应多个音则只重复加入，第3项才是分音节加入，可以自行查看
- 第四维可以说是音节啥的，一个音对应一个字的就一个元素，其他的可能有几个元素。
> 比如在一个歌曲里，一个字“好”对应了两个音，
> 则第三维度和第四维度就表示:[...[“好”,“好”]...]，
> 那么midi则记录为:[...[[pitch,time,restTime],[pitch,time,restTime]]...]
> 所以midi有五维

### 数据使用
接下来则是稍微改一下那个<u>https://github.com/microsoft/muzic/blob/main/songmass/data/generate_lmd_dataset.py</u> 的第109行的这几个代码，因为我们只有0和1，没有1和2

In [326]:
i=0
for word,midi in zip(wordOrigin,midiOrigin):
    data=[word,midi]
    d=np.array([data])
    print(d.shape)
    np.save(os.path.join(lmd_output_path,nameData[i][:-4]+'npy'), d)
    i+=1

C:\Users\86147\AppData\Local\Temp\ipykernel_4164\3730831908.py:4: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  d=np.array([data])


(1, 2, 1, 49)
(1, 2, 8)
(1, 2, 9)
(1, 2, 10)
(1, 2, 10)
(1, 2, 8)
(1, 2, 8)
(1, 2, 13)
(1, 2, 20)
(1, 2, 17)
(1, 2, 1, 71)
(1, 2, 10)
(1, 2, 20)
(1, 2, 7)
(1, 2, 1, 43)
(1, 2, 8)
(1, 2, 9)
(1, 2, 10)
(1, 2, 10)
(1, 2, 4)
(1, 2, 3)
(1, 2, 20)
(1, 2, 7)
(1, 2, 8)
(1, 2, 7)
(1, 2, 8)
(1, 2, 8)
(1, 2, 8)
(1, 2, 14)
(1, 2, 14)
(1, 2, 10)
(1, 2, 5)
(1, 2, 4)
(1, 2, 7)
(1, 2, 5)
(1, 2, 4)
(1, 2, 6)
(1, 2, 9)
(1, 2, 10)
(1, 2, 10)
(1, 2, 11)
(1, 2, 10)
(1, 2, 11)
(1, 2, 9)
(1, 2, 8)
(1, 2, 8)
(1, 2, 8)
(1, 2, 8)
(1, 2, 9)
(1, 2, 9)
(1, 2, 8)
(1, 2, 8)
(1, 2, 8)
(1, 2, 8)
(1, 2, 10)
(1, 2, 8)
(1, 2, 9)
(1, 2, 9)
(1, 2, 9)
(1, 2, 8)
(1, 2, 8)
(1, 2, 6)
(1, 2, 8)
(1, 2, 10)
(1, 2, 10)
(1, 2, 11)
(1, 2, 8)
(1, 2, 18)
(1, 2, 6)
(1, 2, 7)
(1, 2, 10)
(1, 2, 5)
(1, 2, 4)
(1, 2, 8)
(1, 2, 11)
(1, 2, 11)
(1, 2, 21)
(1, 2, 18)
(1, 2, 7)
(1, 2, 10)
(1, 2, 13)
(1, 2, 11)
(1, 2, 8)
(1, 2, 7)
(1, 2, 10)
(1, 2, 5)
(1, 2, 8)
(1, 2, 9)
(1, 2, 9)
(1, 2, 18)
(1, 2, 6)
(1, 2, 13)
(1, 2, 9)
(1, 2, 9)
(1, 2, 8)
(1, 

### 结果：
输出在Data/GeneratedNpy文件夹内
结构与imd文件结构一致

接下来运行下面代码，也就是<u>https://github.com/microsoft/muzic/blob/main/songmass/data/generate_lmd_dataset.py</u>代码，我把它抄下来了

> 不会用，在jupyter里运行了之后报错，在miniconda里也运行了，还是报错，运行结果在Output文件夹中，全是空文件，你们看看如何使用

In [327]:
import os
import io
import argparse
import copy

import numpy as np
from collections import Counter
from nltk.tokenize import word_tokenize


parser = argparse.ArgumentParser(description="LMD data generation")

parser.add_argument('--lmd-data-dir', type=str, default="Data/GeneratedNpy/",
                    help="Lmd data directory")
parser.add_argument('--output-dir', type=str, default="Data/SongMassOutput/",
                    help="Output directory")


REST_NOTE, SEP, ALIGN = 128.0, '[sep]', '[align]'


duration_vocab = dict([
    (str(x / 100), 129 + i) for i, x in enumerate(list(range(25, 3325, 25)))
])


base_tones = {
    'C' : 0, 'C#': 1, 'D' : 2, 'D#': 3,
    'E' : 4, 'F' : 5, 'F#': 6, 'G' : 7,
    'G#': 8, 'A' : 9, 'A#':10, 'B' :11,
}


all_scales = {
    'major'         : [0, 2, 4, 5, 7, 9, 11],
    'natural_minor' : [0, 2, 3, 5, 7, 8, 10],
    'harmonic_minor': [0, 2, 3, 5, 7, 8, 11],
    'melodic_minor' : [0, 2, 3, 5, 7, 9, 11],
}


def tones_to_scales(tones):
    """
      Midi to tone name (octave: -5):
        0  1  2  3  4  5  6  7  8  9  10  11
        C  C# D  D# E  F  F# G  G# A  A#  B
        
      Melodic minor scale is ignored. One octave is 12 tones.
    """
    counts = {
        tone: {
            scale: 0 for scale in all_scales
        }
        for tone in base_tones
    }

    if not tones:
        return None
    
    for cur_tone_value in tones:
        for base_tone_name, base_tone_value in base_tones.items():
            for scale_name, scale_notes in all_scales.items():
                if (cur_tone_value - base_tone_value) % 12 in scale_notes:
                    counts[base_tone_name][scale_name] += 1

    frequency = {}
    for base_tone in counts:
        frequency[base_tone] = {}
        for scale_label in counts[base_tone]:
            frequency[base_tone][scale_label] = counts[base_tone][scale_label] / len(tones)
    return frequency


def maximum_likelihood_scale(tones):
    statistic_scales = tones_to_scales(tones)
    a = []
    for base_tone in statistic_scales:
        for scale_label in statistic_scales[base_tone]:
            a.append((base_tone, scale_label, statistic_scales[base_tone][scale_label]))
    a.sort(key=lambda v: v[2], reverse=True)
    return a[0][0], a[0][1], a[0][2]


def normalize_song(tones):
    base_tone, scale_type, scale_score = maximum_likelihood_scale([t for t in tones if t != REST_NOTE])

    if 'major' in scale_type:
        offset = base_tones[base_tone] - base_tones['C']
    else:
        assert 'minor' in scale_type
        offset = base_tones[base_tone] - base_tones['A']

    tones = [t - offset if t != REST_NOTE else REST_NOTE for t in tones]

    most_common_octave = Counter([t // 12 for t in tones]).most_common(1)[0][0]
    offset = 12 * (most_common_octave - 5)

    tones = [t - offset if t != REST_NOTE else REST_NOTE for t in tones]
    return tones


def extract_melody_and_lyric(file_path):
    data = np.load(file_path, allow_pickle=True)[0]
    raw_melody, raw_lyric = data[0], data[1]

    whole_melody, whole_lyric = [], []
    tmp_melody, tmp_lyric = [], []

    # check file is empty or not, and filter empty sentences
    for melody, lyric in zip(raw_melody, raw_lyric):
        if len(melody) > 0 and len(lyric) > 0:
            whole_melody.extend(melody)
            whole_lyric.extend(lyric)

            tmp_melody.append(melody)
            tmp_lyric.append(lyric)
    
    raw_melody = tmp_melody
    raw_lyric = tmp_lyric
    
    if len(whole_melody) == 0 or len(whole_lyric) == 0:
        return [], []

    def generate_sentence(s):
        word_list = []

        for w in s:
            word_list.append(w[0])
            word_list.append(ALIGN)

        word_list.append(SEP)
        return word_list

    lyric = []
    for x in raw_lyric:
        lyric.extend(generate_sentence(x))

    all_notes = []
    for sent in raw_melody:
        for word in sent:
            for syllable in word:
                all_notes.append(syllable[0])
    normalized_notes = normalize_song(all_notes)

    normalized_melody = copy.deepcopy(raw_melody)
    idx = 0
    for i, melody in enumerate(raw_melody):
        for j, word in enumerate(melody):
            for k, syllable in enumerate(word):
                normalized_melody[i][j][k][0] = normalized_notes[idx]
                idx = idx + 1
    
    notes, durations = [], []

    for sent in normalized_melody:
        cur_sent_notes = []
        cur_sent_durations = []
        cur_sent = []
        for word in sent:
            for syllable in word:
                if syllable[2] > 0:
                    pitch = REST_NOTE
                    duration = round(syllable[2] / 0.25) * 0.25
                    cur_sent_notes.append(pitch)
                    cur_sent_durations.append(duration)

                pitch = syllable[0]
                duration = round(syllable[1] / 0.25) * 0.25
                cur_sent_notes.append(pitch)
                cur_sent_durations.append(duration)
            
            cur_sent_notes.append(REST_NOTE) # word-level alignment flag
            cur_sent_durations.append(-2)

        notes.extend(cur_sent_notes)
        durations.extend(cur_sent_durations)

        notes.append(REST_NOTE)
        durations.append(-1) # sentence-level alignment flag

    durations = [min(x, 4.0) for x in durations]
    melody = list(zip(notes, durations))

    return melody, lyric


def check_length(x):
    pitchs = [int(i[0]) for i in x]
    pitchs = list(filter(lambda x : x != REST_NOTE, pitchs))

    cur_num, cur_cnt, max_repeat = -1, -1, 6

    for i in pitchs:
        if i != cur_num:
            if cur_cnt >= max_repeat:
                return False
            cur_num = i
            cur_cnt = 1
        else:
            cur_cnt += 1
    return True


def generate_pitch_duration_sequence(notes):
    ret_list = []
    for x in notes:
        if x[1] == -2:
            ret_list.append(ALIGN)
        elif x[1] == -1:
            ret_list.append(SEP)
        else:
            ret_list.append(str(int(x[0])))
            ret_list.append(str(duration_vocab[str(x[1])]))
    return ret_list


def find_forward(melody_idx, lyric_idx, start, length=250):
    i = start
    for i in range(start + 1, len(melody_idx)):
        if melody_idx[i] - melody_idx[start] > length or lyric_idx[i] - lyric_idx[start] > length:
            i -= 1
            break
    if i == start:
        i += 1
    return i


def find_backward(melody_idx, lyric_idx, start, length=500):
    i = start
    for i in range(start - 1, -1, -1):
        if melody_idx[start] - melody_idx[i] > length or lyric_idx[start] - lyric_idx[i] > length:
            i -= 1
            break
    if i == start:
        i -= 1
    return i


def sliding_window(melody, lyric, max_stride_size=250, max_window_size=500):
    ret_melody, ret_lyric = [], []
    melody_idx = [i for i, x in enumerate(melody) if x == SEP]
    lyric_idx = [i for i, x in enumerate(lyric) if x == SEP]

    assert len(melody_idx) == len(lyric_idx)
    melody_idx.insert(0, -1)
    lyric_idx.insert(0, -1)

    head = 0
    tail = find_forward(melody_idx, lyric_idx, head, max_window_size)
    ret_melody.append(melody[melody_idx[head] + 1 : melody_idx[tail] + 1])
    ret_lyric.append(lyric[lyric_idx[head] + 1 : lyric_idx[tail] + 1])

    while tail != len(melody_idx) - 1:
        tail = find_forward(melody_idx, lyric_idx, tail, max_stride_size)
        head = find_backward(melody_idx, lyric_idx, tail, max_window_size)
        ret_melody.append(melody[melody_idx[head] + 1 : melody_idx[tail] + 1])
        ret_lyric.append(lyric[lyric_idx[head] + 1 : lyric_idx[tail] + 1])
    return ret_melody, ret_lyric


def cut_window(melody, lyric, max_window_size=500):
    ret_melody = []
    ret_lyric = []

    melody_idx = [i for i, x in enumerate(melody) if x == SEP]
    lyric_idx = [i for i, x in enumerate(lyric) if x == SEP]

    assert len(melody_idx) == len(lyric_idx)
    melody_idx.insert(0, -1)
    lyric_idx.insert(0, -1)

    head, tail = 0, 0
    while tail <= len(lyric_idx):
        if tail == len(lyric_idx) and head < tail:
            tail -= 1
            ret_melody.append(melody[melody_idx[head] + 1 : melody_idx[tail] + 1])
            ret_lyric.append(lyric[lyric_idx[head] + 1 : lyric_idx[tail] + 1])
            break
        
        if melody_idx[tail] - melody_idx[head] > max_window_size or lyric_idx[tail] - lyric_idx[head] > max_window_size:
            tail -= 1
            ret_melody.append(melody[melody_idx[head] + 1 : melody_idx[tail] + 1])
            ret_lyric.append(lyric[lyric_idx[head] + 1 : lyric_idx[tail] + 1])
            head = tail

        tail += 1

    return ret_melody, ret_lyric


def lower_fn(x):
    return " ".join(x).lower() + "\n"


def main(args):
    melody_list, lyric_list = [], []
    for file_name in os.listdir(args.lmd_data_dir):
        file_path = os.path.join(args.lmd_data_dir, file_name)
        melody, lyric = extract_melody_and_lyric(file_path)
        
        if len(melody) == 0 or len(lyric) == 0:
            continue

        if not check_length(melody):
            continue

        melody = generate_pitch_duration_sequence(melody)

        melody_list.append(melody)
        lyric_list.append(lyric)

    n = len(melody_list)

    melody_train = melody_list[int(n * 0.2) : ]
    melody_valid = melody_list[int(n * 0.1) : int(n * 0.2)]
    melody_test  = melody_list[int(n * 0.0) : int(n * 0.1)]

    lyric_train = lyric_list[int(n * 0.2) : ]
    lyric_valid = lyric_list[int(n * 0.1) : int(n * 0.2)]
    lyric_test  = lyric_list[int(n * 0.0) : int(n * 0.1)]
    
    # The training data adopts sliding window, 
    # while the valid/test data directly will be performed by clip-off.
    
    sliding_melody_train = []
    sliding_lyric_train = []
    for melody, lyric in zip(melody_train, lyric_train):
        m, l = sliding_window(melody, lyric)
        sliding_melody_train.extend(m)
        sliding_lyric_train.extend(l)

    cut_melody_valid, cut_lyric_valid, song_id_valid = [], [], []
    for i, (melody, lyric) in enumerate(zip(melody_valid, lyric_valid)):
        m, l = cut_window(melody, lyric)
        cut_melody_valid.extend(m)
        cut_lyric_valid.extend(l)
        song_id_valid.extend([i] * len(m))

    cut_melody_test, cut_lyric_test, song_id_test = [], [], []
    for i, (melody, lyric) in enumerate(zip(melody_test, lyric_test)):
        m, l = cut_window(melody, lyric)
        cut_melody_test.extend(m)
        cut_lyric_test.extend(l)
        song_id_test.extend([i] * len(m))

    
    output_dir = args.output_dir
    output_para_dir = os.path.join(output_dir, 'para')

    if not os.path.exists(output_para_dir):
        os.makedirs(output_para_dir)
    
    with open(os.path.join(output_para_dir, 'train.lyric'), 'w') as f:
        lines = list(map(lower_fn, sliding_lyric_train))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'train.melody'), 'w') as f:
        lines = list(map(lower_fn, sliding_melody_train))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'valid.lyric'), 'w') as f:
        lines = list(map(lower_fn, cut_lyric_valid))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'valid.melody'), 'w') as f:
        lines = list(map(lower_fn, cut_melody_valid))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'test.lyric'), 'w') as f:
        lines = list(map(lower_fn, cut_lyric_test))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'test.melody'), 'w') as f:
        lines = list(map(lower_fn, cut_melody_test))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'song_id_valid.txt'), 'w') as f:
        lines = list(map(lambda x : str(x) + '\n', song_id_valid))
        f.writelines(lines)

    with open(os.path.join(output_para_dir, 'song_id_test.txt'), 'w') as f:
        lines = list(map(lambda x : str(x) + '\n', song_id_test))
        f.writelines(lines)


if __name__ == '__main__':
    args = parser.parse_args()
    main(args)

usage: ipykernel_launcher.py [-h] [--lmd-data-dir LMD_DATA_DIR]
                             [--output-dir OUTPUT_DIR]
ipykernel_launcher.py: error: unrecognized arguments: --ip=127.0.0.1 --stdin=9014 --control=9012 --hb=9011 --Session.signature_scheme="hmac-sha256" --Session.key=b"6eecf41d-99f9-41c6-982c-facac389bb29" --shell=9013 --transport="tcp" --iopub=9015 --f=c:\Users\86147\AppData\Roaming\jupyter\runtime\kernel-v2-12296T23tUdNxGgY7.json


SystemExit: 2

c:\Users\86147\Miniconda3\envs\nlp2022\lib\site-packages\IPython\core\interactiveshell.py:3369: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)
